### PR TITLE
ARROW-176 Nested extension objects are not handled in auto schema

### PR DESCRIPTION
**Pull request by NoahStapp**

### CODE REVIEW

**Comment by NoahStapp** [comment]
Pre-commit hooks are failing on the PR but are succeeding locally: 

```
(pymongoarrow) nstapp@M-THWWK302GC python % pre-commit run --all-files
check for added large files..............................................Passed
check for case conflicts.................................................Passed
check toml...............................................................Passed
check yaml...............................................................Passed
debug statements (python)................................................Passed
fix end of files.........................................................Passed
forbid new submodules................................(no files to check)Skipped
trim trailing whitespace.................................................Passed
black....................................................................Passed
isort....................................................................Passed
flake8...................................................................Passed
shellcheck...............................................................Passed
Check GitHub Workflows...................................................Passed
blacken-docs.............................................................Passed
```

**Comment by NoahStapp** [comment]
> 

Still the same result after uninstalling pre-commit, removing the cache, and re-installing:

```
(pymongoarrow) nstapp@M-THWWK302GC python % pre-commit run --all-files
check for added large files..............................................Passed
check for case conflicts.................................................Passed
check toml...............................................................Passed
check yaml...............................................................Passed
debug statements (python)................................................Passed
fix end of files.........................................................Passed
forbid new submodules................................(no files to check)Skipped
trim trailing whitespace.................................................Passed
black....................................................................Passed
isort....................................................................Passed
flake8...................................................................Passed
shellcheck...............................................................Passed
Check GitHub Workflows...................................................Passed
blacken-docs.............................................................Passed
```

File: bindings/python/pymongoarrow/lib.pyx [hunk 1]
```diff
@@ -781,7 +783,30 @@ cdef class DocumentBuilder(_ArrayBuilderBase):
...
             self.builder.get().Finish(&out)
-        return pyarrow_wrap_array(out)
+
+        struct_array = pyarrow_wrap_array(out)
+        for struct_def in struct_array:
+            new_types = []
+            new_names = list(struct_def.keys())
+            for fname, ftype in struct_def.items():
+                builder_name = type(self.builder_map[fname.encode('utf-8')]).__name__
+                if builder_name == ObjectIdBuilder.__name__: # ObjectIdType
```
[the remark this answers is not in the document]
Comment (NoahStapp) [reply bindings/python/pymongoarrow/lib.pyx:793]: Using `isinstance` is likely better, good point.

**Comment by ShaneHarvey** [comment]
Are the benchmark failures due to the changes here or are they just false positives (caused by perf noise in the runner)?

**Comment by NoahStapp** [comment]
> Are the benchmark failures due to the changes here or are they just false positives (caused by perf noise in the runner)?

It's possible they're caused by the changes here as the benchmarks experiencing regressions are `Document` tests, but I'm not familiar enough with the benchmarks to conclusively say so.

**Comment by ShaneHarvey** [comment]
@blink1073 could it be we're copying more data around than we need to when doing this new casting?

**Comment by blink1073** [comment]
Yeah, a ratio of `2.31` on `benchmarks.ProfileReadDocument.time_to_arrow` seems to show Arrow has to recreate the full document.

**Comment by blink1073** [comment]
What we have here is more technically correct though, so I think it is worth opening a new ticket to track using custom C builder classes in a minor release as opposed to this casting after the fact.

**Comment by ShaneHarvey** [comment]
SGTM!

**Review by ShaneHarvey — APPROVED** [review]

**Comment by NoahStapp** [comment]
> What we have here is more technically correct though, so I think it is worth opening a new ticket to track using custom C builder classes in a minor release as opposed to this casting after the fact.

So this change is fine for now until we implement this same behavior with custom builder classe>

**Comment by blink1073** [comment]
> So this change is fine for now until we implement this same behavior with custom builder classes?

Yes

**Comment by blink1073** [comment]
Can you please revert the import sorting changes that are failing the pre-commit build?  You may have to do it in the GitHub UI (hit the period key to open up the editor).

**Review by blink1073 — APPROVED** [review]
Thanks!

**Pull request merged**